File: nim/run-length-encoding/run_length_encoding.nim
import parseutils, strutils

proc encode*(plainText: string): string =
  var i = 0

  while i < plainText.len:
    let letter = plainText[i]
    let count = skipWhile(plainText, { letter }, i)
    i.inc count

    result.add(if count == 1: $letter else: $count & letter)

proc decode*(cipherText: string): string =
  var i = 0

  while i < cipherText.len:
    var count = 1
    i.inc parseInt(cipherText, count, i)
    
    result.add(cipherText[i].repeat(count))
    i.inc

File: kotlin/run-length-encoding/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {
    fun encode(input: String): String =
        input.encodeRuns().joinToString("") { (letter, count) -> letter.encode(count) }

    fun decode(input: String): String =
        input.decodeRuns().joinToString("") { (letter, count) -> letter.decode(count) }

    private fun Char.encode(count: Int) = if (count == 1) "$this" else "$count$this"

    private fun Char.decode(count: Int) = this.toString().repeat(count)

    private fun String.encodeRuns() =
        Regex("(.)\\1*").findAll(this).map { it.value[0] to it.value.length }

    private fun String.decodeRuns() =
        Regex("(\\d*)(.)").findAll(this).map { it.groupValues[2][0] to it.groupValues[1].ifEmpty { "1" }.toInt() }
}
